{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab9230ad",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "## Get 50.000 mags from AUs study, on zenodo\n",
    "mamba activate zenodo_get\n",
    "\n",
    "zenodo_get -h\n",
    "\n",
    "https://zenodo.org/api/records/17252731\n",
    "https://zenodo.org/records/17250845"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ed8bf2",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# unzip \n",
    "tar -xvzf your_file.tar.gz -C /path/to/destination/folder\n",
    "\n",
    "for f in *mags.tar.gz \n",
    "do\n",
    "tar -xvzf $f -C ../fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eee53cc",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# in: /home/ctbrown/scratch3/2025-other-pig-bins/annie-dl/\n",
    "# make manysketch file\n",
    "\n",
    "# sketch them\n",
    "echo name,genome_filename,protein_filename > 251007_manysketch_Gaio2021.csv\n",
    "for i in */*.fa\n",
    "do\n",
    "  echo \"$i,$(realpath \"$i\"),\"\n",
    "done >> /home/ctbrown/scratch3/2025-other-pig-bins/annie-dl/fasta/251007_manysketch_Gaio2021.csv\n",
    "\n",
    "\n",
    "# then for sketching\n",
    "srun --account=ctbrowngrp -p med2 -J sketch -t 6:00:00 -c 60 --mem=80gb --pty bash\n",
    "\n",
    "# sketch\n",
    "mamba activate branchwater-skipmer\n",
    "sourmash scripts manysketch -p dna,k=21,k=31,k=51,scaled=1000,abund \\\n",
    "-c 60 251007_manysketch_Gaio2021.csv -o MAGs_Gaio2021.all.zip\n",
    "\n",
    "\n",
    "# then index them for a rocksdb \n",
    "sourmash index MAGs_Gaio2021.all.rocksdb -F rocksdb -k 31 MAGs_Gaio2021.all.zip\n",
    "\n",
    "# then also make a pangenome db bc its faster \n",
    "mamba activate pangenomics_dev\n",
    "sourmash scripts pangenome_merge \\\n",
    "MAGs_Gaio2021.all.zip -k 31 --scaled 1000 -o MAGs_Gaio2021.merged.k31.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9ed66da",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# tax for uncassified\n",
    "# /home/ctbrown/scratch3/sourmash-midgie-raker/outputs.dg/rename/raw-sketch.unclassified.csv\n",
    "# bins still in: /home/ctbrown/scratch3/2025-other-pig-bins/annie-dl/\n",
    "\n",
    "srun --account=ctbrowngrp -p med2 -J gtdbtk -t 62:00:00 -c 52 --mem=120gb --pty bash\n",
    "\n",
    "# do it in folders so that we dont do 300 at once bc the compyter doesnt like\n",
    "mamba activate gtdbtk\n",
    "gtdbtk classify_wf  --pplacer_cpus 16 \\\n",
    "--cpus 52 --batchfile gtdb_tk_btchfile.tsv \\\n",
    "--extension fasta \\\n",
    "--skip_ani_screen --out_dir taxonomy_MAGs_Gaio2021\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
